src/sql/src/query_model/hir/hir_from_qgm.rs

```diff
@@ -24,23 +31,393 @@ impl From<Model> for HirRelationExpr {
+            }
+            _ => unimplemented!(),
+        };
+
+        if r#box.ranging_quantifiers().count() > 1 {
+            let id = mz_expr::LocalId::new(self.id_gen.allocate_id());
+            let typ =
+                mz_repr::RelationType::new(r#box.attributes.get::<BoxRelationType>().to_owned());
+            self.common_subgraphs.insert(r#box.id, self.lets.len());
+            self.lets.push((id, hir, typ));
+        } else {
+            self.converted.push(hir);
+        }
+    }
+
+    /// Convert the quantifiers of `box` into [HirRelationExpr] and [HirScalarExpr].
+    ///
+    /// Results are returned in reverse quantifier order; the last entry in the
+    /// second vector corresponds to the first subquery quantifier in the box.
+    fn convert_quantifiers<'a>(
+        &mut self,
+        r#box: &BoundRef<'a, QueryBox>,
+    ) -> (
+        Vec<(QuantifierId, HirRelationExpr)>,
+        Vec<(QuantifierId, HirScalarExpr)>,
+    ) {
+        let mut rels = Vec::new();
+        let mut scalars = Vec::new();
+        for q in r#box.input_quantifiers().rev() {
+            // Retrive the HIR corresponding to the input box of the quantifier
+            let inner_relation = if let Some(let_pos) = self.common_subgraphs.get(&q.input_box) {
+                if matches!(self.lets[*let_pos].1, HirRelationExpr::Get { .. }) {
+                    self.lets[*let_pos].1.clone()
+                } else {
+                    HirRelationExpr::Get {
+                        id: mz_expr::Id::Local(self.lets[*let_pos].0),
+                        typ: self.lets[*let_pos].2.clone(),
+                    }
+                }
+            } else {
+                self.converted.pop().unwrap()
+            };
+
+            match q.quantifier_type {
+                QuantifierType::Foreach | QuantifierType::PreservedForeach => {
+                    rels.push((q.id, inner_relation))
+                }
+                QuantifierType::Existential => {
+                    scalars.push((q.id, HirScalarExpr::Exists(Box::new(inner_relation))))
+                }
+                QuantifierType::Scalar => {
+                    scalars.push((q.id, HirScalarExpr::Select(Box::new(inner_relation))))
+                }
+                QuantifierType::All => {
+                    // TODO: Shouldn't QuantifierType::Any also exist?
+                    unimplemented!()
+                }
+            }
+        }
+        (rels, scalars)
+    }
+
+    /// Convert the projection of a box into a Map + Project around the
+    /// [HirRelationExpr] representing the rest of the box.
+    fn convert_projection<'a>(
+        &mut self,
+        r#box: &BoundRef<'a, QueryBox>,
+        column_map: &ColumnMap,
+        mut result: HirRelationExpr,
+    ) -> HirRelationExpr {
+        let mut map = Vec::new();
+        let mut project = Vec::new();
+        let mut total_columns = column_map.arity();
```

Comment:
I'm not sure that including the subquery quantifiers in this map will not result in issues. @philip-stoev should probably try to find some when we do end-to-end tests.